oak_crypto/src/kem.rs

```diff
@@ -0,0 +1,134 @@
+//
+// Copyright 2023 The Project Oak Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+use crate::kdf::Kdf;
+use anyhow::{anyhow, Context};
+use core::convert::TryInto;
+use p256::{ecdh::EphemeralSecret, EncodedPoint, PublicKey};
+use rand_core::OsRng;
+
+/// DHKEM(P-256, HKDF-SHA256) ID.
+/// <https://www.rfc-editor.org/rfc/rfc9180.html#name-key-encapsulation-mechanism>
+pub(crate) const KEM_ID: u16 = 0x0010;
+/// Represents `N_secret` from RFC9180.
+/// <https://www.rfc-editor.org/rfc/rfc9180.html#name-cryptographic-dependencies>
+pub(crate) const KEM_SHARED_SECRET_SIZE_BYTES: usize = 32;
+/// Represents `N_pk` from RFC9180.
+/// <https://www.rfc-editor.org/rfc/rfc9180.html#name-cryptographic-dependencies>
+pub(crate) const KEM_ENCODED_PUBLIC_KEY_SIZE_BYTES: usize = 65;
+/// Convenience type for representing a KEM shared secret.
+pub(crate) type SharedSecret = [u8; KEM_SHARED_SECRET_SIZE_BYTES];
+/// Convenience type for representing both encoded and encapsulated KEM public keys.
+/// RFC9180 specifies that `Npk` is equal to `Nenc`.
+/// <https://www.rfc-editor.org/rfc/rfc9180.html#name-dh-based-kem-dhkem>
+pub(crate) type SerializedPublicKey = [u8; KEM_ENCODED_PUBLIC_KEY_SIZE_BYTES];
+
+/// Implementation of the Key Encapsulation Mechanism.
+/// <https://www.rfc-editor.org/rfc/rfc9180.html#name-dh-based-kem-dhkem>
+pub(crate) struct Kem {}
+
+impl Kem {
+    /// Randomly generates a key pair.
+    pub(crate) fn generate_key_pair() -> EphemeralSecret {
+        EphemeralSecret::random(&mut OsRng)
+    }
+
+    pub(crate) fn extract_and_expand(
+        diffie_hellman_key_material: &[u8],
+        kem_context: &[u8],
+    ) -> anyhow::Result<SharedSecret> {
+        // Suite ID value used by the KDF.
+        // <https://www.rfc-editor.org/rfc/rfc9180.html#name-dh-based-kem-dhkem>
+        let kem_suite_id = ["KEM".as_bytes(), &KEM_ID.to_be_bytes()].concat();
```

Comment:
```suggestion
        let kem_suite_id = [b"KEM", &KEM_ID.to_be_bytes()].concat();
```

Reply:
If I use a byte string here, than `concat` won't understand the types of the operands (since they will be array of fixed but different lengths):
```
  --> oak_crypto/src/kem.rs:55:37
   |
55 |         let kem_suite_id = [b"KEM", &KEM_ID.to_be_bytes()].concat();
   |                                     ^^^^^^^^^^^^^^^^^^^^^ expected an array with a fixed size of 3 elements, found one with 2 elements
```